File: ios/EpicRunner/EpicRunner/RunSelectorItemView.m
//
//  RunSelectorItemView.m
//  EpicRunner
//
//  Created by Jeppe on 01/07/14.
//  Copyright (c) 2014 Pandisign ApS. All rights reserved.
//

#import "RunSelectorItemView.h"

@interface RunSelectorItemView ()
    @property CGPoint originalLoc;
    @property CGPoint currentLoc;
    @property BOOL toBeDeleted;
    @property UILabel *lblTimeBar;
@end


@implementation RunSelectorItemView
- (id)initWithFrame:(CGRect)frame andText:(NSString*)text
{
    self = [super initWithFrame:frame];
    if (self) {
        self.userInteractionEnabled = YES;
        self.originalLoc = self.frame.origin;
        self.backgroundColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0];
        self.toBeDeleted = NO;
        
        // Register tap gesture
        UITapGestureRecognizer *singleFingerTap =
        [[UITapGestureRecognizer alloc] initWithTarget:self
                                                action:@selector(handleSingleTap:)];
        [self addGestureRecognizer:singleFingerTap];
        
        // Register long tap
        UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc]
                                                   initWithTarget:self
                                                   action:@selector(handleLongPress:)];
        longPress.minimumPressDuration = 1.5;
        [self addGestureRecognizer:longPress];
        
        // Create timebar label
        self.lblTimeBar = [[UILabel alloc] initWithFrame: CGRectMake(0,38,0,3)];
        UIColor *color = [UIColor colorWithRed:0.2 green:1.0 blue:0.2 alpha:1.0];
        self.lblTimeBar.backgroundColor = color;
        [self addSubview:self.lblTimeBar];
    
        // Insert Text
        UILabel *label = [[UILabel alloc] initWithFrame: CGRectMake(10,0,100,self.frame.size.height)];
        label.text = text;
        [self addSubview:label];
    }
    return self;
}


// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
//- (void)drawRect:(CGRect)rect
//{
    // Create time-rect
    //CGContextRef context = UIGraphicsGetCurrentContext();
    //UIColor *color = [UIColor colorWithRed:0.2 green:1.0 blue:0.2 alpha:1.0];
    //CGContextSetFillColorWithColor(context, color.CGColor);
    //CGContextFillRect(context, CGRectMake(0,38,30,3));
//}


- (void)handleSingleTap:(UITapGestureRecognizer *)recognizer {
    //CGPoint location = [recognizer locationInView:[recognizer.view superview]];
    NSLog(@"TAPPED");
}

- (void)handleLongPress:(UILongPressGestureRecognizer*)recognizer {
    CGPoint point = [recognizer locationInView:self.superview];
    
    if (recognizer.state == UIGestureRecognizerStateChanged) {
        // Move to touched position
        CGPoint center = self.center;
        center.x += point.x - self.currentLoc.x;
        center.y += point.y - self.currentLoc.y;
        self.center = center;
        
        // Check for dropzone
        //NSLog(@"%.f", self.frame.origin.y);
        if (self.frame.origin.y > 320 && self.frame.origin.y < 345) {
            self.toBeDeleted = YES;
            self.backgroundColor = [UIColor colorWithRed:0.7 green:0.1 blue:0.1 alpha:1.0];
        } else {
            self.toBeDeleted = NO;
            self.backgroundColor = [UIColor colorWithRed:0.8 green:0.8 blue:0.8 alpha:1.0];
        }
    }
    else if (recognizer.state == UIGestureRecognizerStateBegan){
        NSLog(@"LONG PRESS BEGAN");
        
        self.backgroundColor = [UIColor colorWithRed:0.8 green:0.8 blue:0.8 alpha:1.0];
    }
    else if (recognizer.state == UIGestureRecognizerStateEnded){
        NSLog(@"LONG PRESS ENDED");
        
        self.backgroundColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0];
        
        // Check for dropzone
        if (self.toBeDeleted == YES) {
            [self destroy];
        } else {
            //  Restore position
            [UIView animateWithDuration:0.5
                                  delay:0.1
                                options: (UIViewAnimationOptions)UIViewAnimationCurveEaseOut
                             animations:^{
                                 CGRect frame = self.frame;
                                 frame.origin = self.originalLoc;
                                 self.frame = frame;
                             }
                             completion:^(BOOL finished) {
                                 NSLog(@"Completed");
                             }];
        }
    }
    
    self.currentLoc = point;
}

- (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
    NSLog(@"touch begin");
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    NSLog(@"touch ended");
}

- (void)moveRelative:(BOOL)relative coordX:(int)x coordY:(int)y {
    [UIView animateWithDuration:0.5
                          delay:0.1
                        options: (UIViewAnimationOptions)UIViewAnimationCurveEaseOut
                     animations:^{
                         CGRect frame = self.frame;
                         if (relative) {
                             frame.origin.x += x;
                             frame.origin.y += y;
                         } else {
                             frame.origin.x = x;
                             frame.origin.y = y;
                         }
                         self.originalLoc = frame.origin;
                         self.frame = frame;
                     }
                     completion:^(BOOL finished) {
                         NSLog(@"Completed");
                     }];
}

- (BOOL)tick {
    int i = 15;

    if (self.lblTimeBar.frame.size.width + i > self.frame.size.width) {
        // 100 %, we're done
        [self.lblTimeBar setFrame:CGRectMake(self.lblTimeBar.frame.origin.x,
                                             self.lblTimeBar.frame.origin.y,
                                             self.frame.size.width,
                                             self.lblTimeBar.frame.size.height)];
        
        UIColor *color = [UIColor colorWithRed:1.0 green:0.2 blue:0.2 alpha:1.0];
        self.lblTimeBar.backgroundColor = color;

        NSLog(@"My time is up on this planet! [item commits suicide]");

        dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, 3 * NSEC_PER_SEC);
        dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
            // Execute code on the main queue after delay
            [self destroy];
        });
        return YES;
    } else if (self.lblTimeBar.frame.size.width/self.frame.size.width > 0.6){
        // Above 60 %, make yellow, add progress
        UIColor *color = [UIColor colorWithRed:1.0 green:1.0 blue:0.2 alpha:1.0];
        self.lblTimeBar.backgroundColor = color;
        
        [self.lblTimeBar setFrame:CGRectMake(self.lblTimeBar.frame.origin.x,
                                             self.lblTimeBar.frame.origin.y,
                                             self.lblTimeBar.frame.size.width+i,
                                             self.lblTimeBar.frame.size.height)];
    } else {
        // All fine, add progress
        [self.lblTimeBar setFrame:CGRectMake(self.lblTimeBar.frame.origin.x,
                                             self.lblTimeBar.frame.origin.y,
                                             self.lblTimeBar.frame.size.width+i,
                                             self.lblTimeBar.frame.size.height)];
    }
    //NSLog(@"ticktick %f", self.lblTimeBar.frame.size.width);
    return NO;
}

- (void)destroy {
    [self removeFromSuperview];
}

@end

File: ios/EpicRunner/EpicRunner/RunSelectorItemView.swift
//
//  RunSelectorItemView.swift
//  EpicRunner
//
//  Created by Jeppe on 15/07/14.
//  Copyright (c) 2014 Pandisign ApS. All rights reserved.
//

import UIKit

class RunSelectorItemView: UIView {
    var originalLoc: CGPoint = CGPoint();
    var currentLoc: CGPoint = CGPoint();
    var toBeLocked: Bool = false;
    var btnTimeBar: UIButton = UIButton(frame: CGRectMake(0, 38, 0, 3));
    let secBetweenTicks = 1.0;
    var tickTimer: NSTimer?;
    let startDate: NSDate;
    let endDate: NSDate;
    let totalTime: NSTimeInterval;
    var destroyed = false;
    var locked = false;
    var runId: Int = 0;
    let lockPlaceholderY: Int = 364;
    let lockPlaceholderX: Int = 20;
    var disabled = false;
    
    init(frame: CGRect, runId: Int, type: String, startDate: NSDate, endDate: NSDate, difficulty: Int, distance: Int, disabled: Bool) {
        self.startDate = startDate;
        self.endDate = endDate;
        self.totalTime = endDate.timeIntervalSince1970 - startDate.timeIntervalSince1970;
        self.disabled = disabled;
        
        super.init(frame: frame);
        
        // Initialization code
        self.originalLoc = self.frame.origin;
        self.userInteractionEnabled = true;
        self.backgroundColor = UIColor(red: 0.9, green: 0.9, blue: 0.9, alpha: 1.0);
        self.runId = runId
        
        var textColor = UIColor(red: 0.0, green: 0.0, blue: 0.0, alpha: 1.0);
        if (disabled) {
            textColor = UIColor(red: 0.333, green: 0.333, blue: 0.333, alpha: 1.0);
        }
        
        // Add timebar
        self.btnTimeBar.contentMode = UIViewContentMode.ScaleToFill;
        self.addSubview(self.btnTimeBar);
        
        // Add type
        let lblType = UILabel(frame: CGRectMake(10, 0, 150, self.frame.size.height));
        lblType.text = type;
        lblType.textColor = textColor;
        self.addSubview(lblType);
        
        // Add disabled
        if (disabled) {
            let lblDisabled = UILabel(frame: CGRectMake(self.frame.size.width-40, 2, 40, 10));
            lblDisabled.text = "Disabled";
            lblDisabled.font = UIFont(name: lblDisabled.font.fontName, size: 8);
            lblDisabled.alignmentRectInsets().right;
            lblDisabled.textColor = textColor;
            self.addSubview(lblDisabled);
            
            self.backgroundColor = UIColor(red: 0.67, green: 0.67, blue: 0.67, alpha: 1.0);
        }
        
        // Add difficulty
        let lblDiff = UILabel(frame: CGRectMake(self.frame.size.width-30, self.frame.size.height/2-7, 14, 14));
        if (difficulty == 1) {
            lblDiff.backgroundColor = UIColor(red: 1.0, green: 0.2, blue: 0.2, alpha: 1.0);
        } else if (difficulty == 2) {
            lblDiff.backgroundColor = UIColor(red: 1.0, green: 1.0, blue: 0.2, alpha: 1.0);
        } else if (difficulty == 3) {
            lblDiff.backgroundColor = UIColor(red: 0.2, green: 1.0, blue: 0.2, alpha: 1.0);
        }
        self.addSubview(lblDiff);
        
        // Add distance
        let lblDist = UILabel(frame: CGRectMake(160, 0, 100, self.frame.size.height));
        lblDist.text = "≈\(distance) km";
        lblDist.textColor = textColor;
        self.addSubview(lblDist);
        
        
        // Register long press
        var longPress = UILongPressGestureRecognizer(target: self, action: "handleLongPress:");
        longPress.minimumPressDuration = 1.0;
        self.addGestureRecognizer(longPress);
        
        // Register tap
        var tap = UITapGestureRecognizer(target: self, action: "handleTap:");
        self.addGestureRecognizer(tap);
        
        
        // Tick
        self.tickTimer = NSTimer.scheduledTimerWithTimeInterval(self.secBetweenTicks, target: self, selector: "tick", userInfo: nil, repeats: true);

    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func handleTap(recognizer: UITapGestureRecognizer) {
        if (self.disabled) {
            // On touch, inform this run has been run
            var alert: UIAlertView = UIAlertView();
            alert.title = "This run is not available";
            alert.message = "This run has already been run and is thus not available anymore.";
            alert.addButtonWithTitle("Ok");
            alert.show();
        } else {
            // On touch open detail view
            let h = superview!.nextResponder() as RunSelectorViewControllerSwift;
            h.selectedRunId = self.runId;
            h.performSegueWithIdentifier("segueDetailView", sender: self);
        }
    }
    
    func move(relative: Bool, coord: CGPoint) {
        UIView.animateWithDuration(0.5,
            delay: 0.1,
            options: .CurveEaseOut,
            animations: { _ in
                if (relative) {
                    self.frame.origin.x += coord.x;
                    self.frame.origin.y += coord.y;
                } else {
                    self.frame.origin.x = coord.x;
                    self.frame.origin.y = coord.y;
                }
                
                // Save current location as "original"
                self.originalLoc = self.frame.origin;
            },
            completion: { _ in
            });
    }

    func handleLongPress(recognizer: UILongPressGestureRecognizer) {
        if (!self.disabled) {
            let point: CGPoint = recognizer.locationInView(self.superview);
            
            switch recognizer.state {
            case UIGestureRecognizerState.Changed:
                // Move to touched position
                self.center.x += point.x - self.currentLoc.x;
                self.center.y += point.y - self.currentLoc.y;
                
                // Check for dropzone
                if (self.frame.origin.y > 345 && self.frame.origin.y < 390) {
                    self.toBeLocked = true;
                    self.backgroundColor = UIColor(red: 0.1, green: 0.7, blue: 0.1, alpha: 1.0);
                } else {
                    self.toBeLocked = false;
                    self.backgroundColor = UIColor(red: 0.8, green: 0.8, blue: 0.8, alpha: 1.0);
                }
                
            case UIGestureRecognizerState.Began:
                println("LONG PRESS BEGAN");
                self.backgroundColor = UIColor(red: 0.8, green: 0.8, blue: 0.8, alpha: 1.0);
                self.layer.zPosition = 99;
                
            case UIGestureRecognizerState.Ended:
                println("LONG PRESS ENDED");
                self.backgroundColor = UIColor(red: 0.9, green: 0.9, blue: 0.9, alpha: 1.0);
                self.layer.zPosition = 1;
                
                // Check for dropzone
                if (self.toBeLocked) {
                    // Lock run
                    lock();
                    
                    // Snap to placeholder
                    UIView.animateWithDuration(0.5,
                        delay: 0.1,
                        options: .CurveEaseOut,
                        animations: { _ in
                            self.frame.origin = CGPoint(x: self.lockPlaceholderX, y: self.lockPlaceholderY);
                        },
                        completion: { _ in
                        }
                    );
                    
                    println("My time among you suckers are over! [item locks]");
                } else {
                    //  Restore position
                    UIView.animateWithDuration(0.5,
                        delay: 0.1,
                        options: .CurveEaseOut,
                        animations: { _ in
                            self.frame.origin = self.originalLoc;
                        },
                        completion: { _ in
                            println("item pos restored");
                        }
                    );
                }
                
            default:
                ();
            }
            
            self.currentLoc = point;
        }
    }
    
    func tick() -> Bool {
    
        
        // In % how close to end date
        let curDate = NSDate().timeIntervalSince1970;
        let t = (curDate - self.startDate.timeIntervalSince1970) / self.totalTime * 100;
        
        if (t > 100) {
            // 100 %, we're done
            println("My time on this planet is over! [item commits suicide]");
            destroy(true);
            
            return false;
            
        } else if (t > 90) {
            // Above 90 %, make red, add progress
            self.btnTimeBar.backgroundColor = UIColor(red: 1.0, green: 0.2, blue: 0.2, alpha: 1.0);
        } else if (t > 75) {
            // Above 75 %, make yellow, add progress
            self.btnTimeBar.backgroundColor = UIColor(red: 1.0, green: 1.0, blue: 0.2, alpha: 1.0);
        } else {
            // All fine, add progress
            self.btnTimeBar.backgroundColor = UIColor(red: 0.2, green: 1.0, blue: 0.2, alpha: 1.0);
        }
        
        
        if (curDate > self.startDate.timeIntervalSince1970) {
            // That % in pixels of the timebars size
            let h = CGFloat(t)
            var p = self.frame.size.width / 100 * h;
            
            if (p > self.frame.size.width) {
                p = self.frame.size.width;
            }
            if (p < 0) {
                p = 0;
            }
            
            UIView.animateWithDuration(self.secBetweenTicks,
                delay: 0.0,
                options: .CurveLinear,
                animations: { _ in
                    self.btnTimeBar.frame.size.width = p;
                },
                completion: { _ in ()}
            );
        }
        
        return false;
    }
    
    func lock() {
        if (!locked) {
            // To avoid the rediculous case where the destroy function already is scheduled
            // and then gets called before the scheduled one
            locked = true;
            
            // Clean up timer (we want no memory leak, yirks!)
            self.tickTimer?.invalidate();
            self.tickTimer = nil;
            
            // Visually remove timer
            UIView.animateWithDuration(self.secBetweenTicks,
                delay: 0.0,
                options: .CurveLinear,
                animations: { _ in
                    self.btnTimeBar.frame.size.width = 0;
                },
                completion: { _ in ()}
            );
            
            // Let the controller know that i'm out
            let h = superview!.nextResponder() as RunSelectorViewControllerSwift;
            h.lockRun(self.runId);
        }
    }
    
    func destroy(timedOut: Bool) {
        if (!self.destroyed) {
            // To avoid the rediculous case where the destroy function already is scheduled
            // and then gets called before the scheduled one
            self.destroyed = true;
            
            //println("destroY: \(self.runId) - \(timedOut)");
            
            // Clean up timer (we want no memory leak, yirks!)
            self.tickTimer?.invalidate();
            self.tickTimer = nil;
            
            // Let the controller know that i'm out
            let h = superview!.nextResponder() as RunSelectorViewControllerSwift;
            h.itemKilled(timedOut);
            
            // Commit suicide
            self.removeFromSuperview();
        }
    }
    
    deinit {
        self.tickTimer?.invalidate();
    }
    
    /*
    // Only override drawRect: if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func drawRect(rect: CGRect)
    {
        // Drawing code
    }
    */

}
